File: source/Game/Fish.lua
class('Fish').extends()

local fishSprites = {}

function Fish:init(position, world)
    self.position = position or Vector3(0, 0, 0)
    self.world = world
    
    -- Load images once
    if #fishSprites == 0 then
        for i = 1, 4 do
            local img = playdate.graphics.image.new('Assets/fish_v' .. i)
            if img then
                local w, h = img:getSize()
                local flipped = playdate.graphics.image.new(w, h)
                playdate.graphics.lockFocus(flipped)
                img:draw(0, 0, playdate.graphics.kImageFlippedX)
                playdate.graphics.unlockFocus()
                
                table.insert(fishSprites, {img = img, flipped = flipped})
            end
        end
    end
    
    -- Assign random sprite
    if #fishSprites > 0 then
        self.spriteIndex = math.random(1, #fishSprites)
    else
        self.spriteIndex = 0
    end
    
    -- Movement
    self.velocity = Vector3(
        math.random() * 2 - 1,
        math.random() * 2 - 1,
        math.random() * 2 - 1
    ):normalize():mul(math.random() * 0.2 + 0.1)
    
    self.speed = math.random() * 0.3 + 0.2
    self.wobblePhase = math.random() * math.pi * 2
    self.turnTimer = math.random() * 3.0 + 2.0 -- Time until direction change
    
    -- Visual
    self.size = math.random() * 0.5 + 0.8 -- Base scale for sprite
    self.rotation = 0 -- Yaw rotation
    self.flip = playdate.graphics.kImageUnflipped

    self.frameCounter = math.random(0, 59)
    self.accumulatedDt = 0
end

function Fish:update(dt)
    local time = playdate.getCurrentTimeMilliseconds() / 1000.0
    
    -- Update AI and collisions only every 1.0 second
    self.accumulatedDt = self.accumulatedDt + dt
    if self.accumulatedDt >= 1.0 then
        -- Update direction periodically
        self.turnTimer = self.turnTimer - self.accumulatedDt
        if self.turnTimer <= 0 then
            self.turnTimer = math.random() * 4.0 + 2.0
            -- Slight direction change
            local turn = Vector3(
                (math.random() * 2 - 1) * 0.3,
                (math.random() * 2 - 1) * 0.3,
                (math.random() * 2 - 1) * 0.3
            )
            self.velocity = self.velocity:add(turn):normalize():mul(self.speed)
        end
        
        -- Wall avoidance
        local limit = self.world.size - 30
        local bounce = 0.5
        
        if self.position.x > limit then
            self.position.x = limit
            self.velocity.x = -math.abs(self.velocity.x) * bounce
        elseif self.position.x < -limit then
            self.position.x = -limit
            self.velocity.x = math.abs(self.velocity.x) * bounce
        end
        
        if self.position.y > limit then
            self.position.y = limit
            self.velocity.y = -math.abs(self.velocity.y) * bounce
        elseif self.position.y < -limit then
            self.position.y = -limit
            self.velocity.y = math.abs(self.velocity.y) * bounce
        end
        
        if self.position.z > limit then
            self.position.z = limit
            self.velocity.z = -math.abs(self.velocity.z) * bounce
        elseif self.position.z < -limit then
            self.position.z = -limit
            self.velocity.z = math.abs(self.velocity.z) * bounce
        end
        
        -- Calculate rotation (yaw) based on velocity direction
        if self.velocity:magnitude() > 0.01 then
            self.rotation = math.atan2(self.velocity.x, self.velocity.z)
            -- Determine flip based on horizontal movement
            if self.velocity.x > 0 then
                self.flip = playdate.graphics.kImageUnflipped -- Right
            else
                self.flip = playdate.graphics.kImageFlippedX -- Left
            end
        end
        
        self.accumulatedDt = 0
    end

    -- Always apply velocity for smooth movement
    self.position = self.position:add(self.velocity:mul(dt * 40))
end

function Fish:draw(camera)
    local proj = camera:project(self.position)
    if proj then
        local spriteData = fishSprites[self.spriteIndex]
        if spriteData then
            -- Scale by depth (proj.scale) and base fish size
            local drawScale = proj.scale * 0.05 * self.size
            
            -- Add a tiny bit of wobble to the sprite position
            local time = playdate.getCurrentTimeMilliseconds() / 1000.0
            local wobbleY = math.sin(time * 5.0 + self.wobblePhase) * 2
            
            local img = spriteData.img
            if self.flip == playdate.graphics.kImageFlippedX then
                img = spriteData.flipped
            end
            
            -- Center the sprite
            local w, h = img:getSize()
            local offX = (w * drawScale) / 2
            local offY = (h * drawScale) / 2
            
            img:drawScaled(proj.x - offX, proj.y + wobbleY - offY, drawScale)
        else
            -- Fallback: simple circle if image is missing
            local gfx = playdate.graphics
            local r = math.max(2, proj.scale * 2 * self.size)
            gfx.setColor(gfx.kColorBlack)
            gfx.fillCircleAtPoint(proj.x, proj.y, r)
        end
    end
end

File: source/Game/Fish.c
#include "Fish.h"

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

static LCDBitmap* fish_sprites[FISH_VARIANTS] = {0};

static float random_range(float min_value, float max_value) {
    const float random_unit = (float)rand() / (float)RAND_MAX;
    return min_value + random_unit * (max_value - min_value);
}

void fish_load_assets(PlaydateAPI* pd) {
    const char* out_error = NULL;
    for (int i = 0; i < FISH_VARIANTS; ++i) {
        if (fish_sprites[i] != NULL) {
            continue;
        }

        char path[32];
        snprintf(path, sizeof(path), "Assets/fish_v%d", i + 1);
        fish_sprites[i] = pd->graphics->loadBitmap(path, &out_error);
    }
}

void fish_init(Fish* fish, Vector3 position, World* world) {
    fish->position = position;
    fish->world = world;

    fish->velocity = vector3_normalize(vector3_create(
        random_range(-1.0f, 1.0f),
        random_range(-1.0f, 1.0f),
        random_range(-1.0f, 1.0f)
    ));

    fish->velocity = vector3_mul(fish->velocity, random_range(0.1f, 0.3f));
    fish->speed = random_range(0.2f, 0.5f);
    fish->wobble_phase = random_range(0.0f, 6.283185f);
    fish->turn_timer = random_range(2.0f, 5.0f);
    fish->size = random_range(0.8f, 1.3f);
    fish->accumulated_dt = 0.0f;
    fish->sprite_index = rand() % FISH_VARIANTS;
}

void fish_update(Fish* fish, float dt) {
    fish->accumulated_dt += dt;

    if (fish->accumulated_dt >= 1.0f) {
        fish->turn_timer -= fish->accumulated_dt;

        if (fish->turn_timer <= 0.0f) {
            fish->turn_timer = random_range(2.0f, 6.0f);

            const Vector3 turn = vector3_create(
                random_range(-0.3f, 0.3f),
                random_range(-0.3f, 0.3f),
                random_range(-0.3f, 0.3f)
            );

            fish->velocity = vector3_add(fish->velocity, turn);
            fish->velocity = vector3_mul(vector3_normalize(fish->velocity), fish->speed);
        }

        const float limit = (float)fish->world->size - 30.0f;
        const float bounce = 0.5f;

        if (fish->position.x > limit) {
            fish->position.x = limit;
            fish->velocity.x = -fabsf(fish->velocity.x) * bounce;
        } else if (fish->position.x < -limit) {
            fish->position.x = -limit;
            fish->velocity.x = fabsf(fish->velocity.x) * bounce;
        }

        if (fish->position.y > limit) {
            fish->position.y = limit;
            fish->velocity.y = -fabsf(fish->velocity.y) * bounce;
        } else if (fish->position.y < -limit) {
            fish->position.y = -limit;
            fish->velocity.y = fabsf(fish->velocity.y) * bounce;
        }

        if (fish->position.z > limit) {
            fish->position.z = limit;
            fish->velocity.z = -fabsf(fish->velocity.z) * bounce;
        } else if (fish->position.z < -limit) {
            fish->position.z = -limit;
            fish->velocity.z = fabsf(fish->velocity.z) * bounce;
        }

        fish->accumulated_dt = 0.0f;
    }

    fish->position = vector3_add(fish->position, vector3_mul(fish->velocity, dt * 40.0f));
}

void fish_draw(const Fish* fish, const Camera* camera, PlaydateAPI* pd) {
    ProjectedPoint projected;
    if (!camera_project(camera, fish->position, &projected)) {
        return;
    }

    const float elapsed_seconds = (float)pd->system->getCurrentTimeMilliseconds() / 1000.0f;
    const float wobble_y = sinf(elapsed_seconds * 5.0f + fish->wobble_phase) * 2.0f;
    const float draw_scale = projected.scale * 0.05f * fish->size;

    LCDBitmap* sprite = fish_sprites[fish->sprite_index];
    if (sprite != NULL && draw_scale > 0.0f) {
        int width;
        int height;
        pd->graphics->getBitmapData(sprite, &width, &height, NULL, NULL, NULL);

        const float offset_x = ((float)width * draw_scale) * 0.5f;
        const float offset_y = ((float)height * draw_scale) * 0.5f;

        pd->graphics->drawScaledBitmap(
            sprite,
            (int)(projected.x - offset_x),
            (int)(projected.y + wobble_y - offset_y),
            draw_scale,
            draw_scale
        );
        return;
    }

    float radius = projected.scale * 2.0f * fish->size;
    if (radius < 2.0f) {
        radius = 2.0f;
    }

    pd->graphics->fillEllipse(
        (int)(projected.x - radius),
        (int)(projected.y - radius),
        (int)(radius * 2.0f),
        (int)(radius * 2.0f),
        0.0f,
        0.0f,
        kColorBlack
    );
}
